File: app/src/main/java/vn/loitp/app/activity/customviews/videoview/exoplayerdemofromggima/PlayerManager.java
/*
 * Copyright (C) 2017 The Android Open Source Project
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *      http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
package vn.loitp.app.activity.customviews.videoview.exoplayerdemofromggima;

import android.content.Context;
import android.net.Uri;
import android.os.Handler;
import android.support.annotation.Nullable;
import android.widget.ImageView;

import com.bumptech.glide.request.target.Target;
import com.github.rubensousa.previewseekbar.base.PreviewLoader;
import com.github.rubensousa.previewseekbar.exoplayer.PreviewTimeBarLayout;
import com.google.ads.interactivemedia.v3.api.player.VideoAdPlayer;
import com.google.android.exoplayer2.C;
import com.google.android.exoplayer2.C.ContentType;
import com.google.android.exoplayer2.ExoPlayer;
import com.google.android.exoplayer2.ExoPlayerFactory;
import com.google.android.exoplayer2.Player;
import com.google.android.exoplayer2.SimpleExoPlayer;
import com.google.android.exoplayer2.ext.ima.ImaAdsLoader;
import com.google.android.exoplayer2.source.ExtractorMediaSource;
import com.google.android.exoplayer2.source.MediaSource;
import com.google.android.exoplayer2.source.MediaSourceEventListener;
import com.google.android.exoplayer2.source.ads.AdsMediaSource;
import com.google.android.exoplayer2.source.dash.DashMediaSource;
import com.google.android.exoplayer2.source.dash.DefaultDashChunkSource;
import com.google.android.exoplayer2.source.hls.HlsMediaSource;
import com.google.android.exoplayer2.source.smoothstreaming.DefaultSsChunkSource;
import com.google.android.exoplayer2.source.smoothstreaming.SsMediaSource;
import com.google.android.exoplayer2.trackselection.AdaptiveTrackSelection;
import com.google.android.exoplayer2.trackselection.DefaultTrackSelector;
import com.google.android.exoplayer2.trackselection.TrackSelection;
import com.google.android.exoplayer2.trackselection.TrackSelector;
import com.google.android.exoplayer2.ui.PlayerView;
import com.google.android.exoplayer2.upstream.BandwidthMeter;
import com.google.android.exoplayer2.upstream.DataSource;
import com.google.android.exoplayer2.upstream.DefaultBandwidthMeter;
import com.google.android.exoplayer2.upstream.DefaultDataSourceFactory;
import com.google.android.exoplayer2.util.Util;

import loitp.basemaster.R;
import vn.loitp.views.uizavideo.glide.GlideApp;
import vn.loitp.views.uizavideo.glide.GlideThumbnailTransformationPB;
import vn.loitp.views.uizavideo.listerner.AudioEventListener;
import vn.loitp.views.uizavideo.listerner.MetadataOutputListener;
import vn.loitp.views.uizavideo.listerner.PlayerEventListener;
import vn.loitp.views.uizavideo.listerner.TextOutputListener;
import vn.loitp.views.uizavideo.listerner.VideoEventListener;
import vn.loitp.core.utilities.LLog;

/**
 * Manages the {@link ExoPlayer}, the IMA plugin and all video playback.
 */
/* package */ final class PlayerManager implements AdsMediaSource.MediaSourceFactory, PreviewLoader {
    private final String TAG = getClass().getSimpleName();
    private final ImaAdsLoader adsLoader;
    private final DataSource.Factory manifestDataSourceFactory;
    private final DataSource.Factory mediaDataSourceFactory;

    private SimpleExoPlayer player;
    private long contentPosition;

    private PreviewTimeBarLayout previewTimeBarLayout;
    private String thumbnailsUrl;
    private ImageView imageView;
    private Player.EventListener eventListener = new Player.DefaultEventListener() {
        @Override
        public void onPlayerStateChanged(boolean playWhenReady, int playbackState) {
            if (playbackState == Player.STATE_READY && playWhenReady) {
                if (previewTimeBarLayout != null) {
                    previewTimeBarLayout.hidePreview();
                }
            }
        }
    };

    public PlayerManager(Context context, PreviewTimeBarLayout previewTimeBarLayout, ImageView imageView, String thumbnailsUrl) {
        String adTag = context.getString(R.string.ad_tag_url);
        adsLoader = new ImaAdsLoader(context, Uri.parse(adTag));
        manifestDataSourceFactory = new DefaultDataSourceFactory(context, Util.getUserAgent(context, context.getString(R.string.app_name)));
        mediaDataSourceFactory = new DefaultDataSourceFactory(
                context,
                Util.getUserAgent(context, context.getString(R.string.app_name)),
                new DefaultBandwidthMeter());
        this.imageView = imageView;
        this.previewTimeBarLayout = previewTimeBarLayout;
        this.thumbnailsUrl = thumbnailsUrl;
    }

    public void init(Context context, PlayerView playerView) {
        // Create a default track selector.
        BandwidthMeter bandwidthMeter = new DefaultBandwidthMeter();
        TrackSelection.Factory videoTrackSelectionFactory = new AdaptiveTrackSelection.Factory(bandwidthMeter);
        TrackSelector trackSelector = new DefaultTrackSelector(videoTrackSelectionFactory);

        // Create a player instance.
        player = ExoPlayerFactory.newSimpleInstance(context, trackSelector);

        // Bind the player to the view.
        playerView.setPlayer(player);

        // This is the MediaSource representing the content media (i.e. not the ad).
        //String contentUrl = context.getString(R.string.content_url);
        String contentUrl = context.getString(R.string.url_dash);
        //String contentUrl = context.getString(R.string.url_hls);
        //String contentUrl = context.getString(R.string.url_smooth);
        MediaSource contentMediaSource = buildMediaSource(Uri.parse(contentUrl), /* handler= */ null, /* listener= */ null);

        // Compose the content media source into a new AdsMediaSource with both ads and content.
        MediaSource mediaSourceWithAds = new AdsMediaSource(
                contentMediaSource,
            /* adMediaSourceFactory= */ this,
                adsLoader,
                playerView.getOverlayFrameLayout(),
            /* eventHandler= */ null,
            /* eventListener= */ null);

        // Prepare the player with the source.
        player.seekTo(contentPosition);

        player.addListener(eventListener);

        player.addListener(new PlayerEventListener());
        player.addAudioDebugListener(new AudioEventListener());
        player.addVideoDebugListener(new VideoEventListener());
        player.addMetadataOutput(new MetadataOutputListener());
        player.addTextOutput(new TextOutputListener());

        adsLoader.addCallback(new VideoAdPlayer.VideoAdPlayerCallback() {
            @Override
            public void onPlay() {
                LLog.d(TAG, "onPlay");
            }

            @Override
            public void onVolumeChanged(int i) {
                LLog.d(TAG, "onVolumeChanged");
            }

            @Override
            public void onPause() {
                LLog.d(TAG, "onPause");
            }

            @Override
            public void onResume() {
                LLog.d(TAG, "onResume");
            }

            @Override
            public void onEnded() {
                LLog.d(TAG, "onEnded");
            }

            @Override
            public void onError() {
                LLog.d(TAG, "onError");
            }
        });

        player.prepare(mediaSourceWithAds);
        player.setPlayWhenReady(true);
    }

    public void reset() {
        if (player != null) {
            contentPosition = player.getContentPosition();
            player.release();
            player = null;
        }
    }

    public void release() {
        if (player != null) {
            player.release();
            player = null;
        }
        adsLoader.release();
    }

    // AdsMediaSource.MediaSourceFactory implementation.

    @Override
    public MediaSource createMediaSource(Uri uri, @Nullable Handler handler, @Nullable MediaSourceEventListener listener) {
        return buildMediaSource(uri, handler, listener);
    }

    @Override
    public int[] getSupportedTypes() {
        // IMA does not support Smooth Streaming ads.
        return new int[]{C.TYPE_DASH, C.TYPE_HLS, C.TYPE_OTHER};
    }

    // Internal methods.

    private MediaSource buildMediaSource(
            Uri uri, @Nullable Handler handler, @Nullable MediaSourceEventListener listener) {
        @ContentType int type = Util.inferContentType(uri);
        switch (type) {
            case C.TYPE_DASH:
                return new DashMediaSource.Factory(
                        new DefaultDashChunkSource.Factory(mediaDataSourceFactory),
                        manifestDataSourceFactory)
                        .createMediaSource(uri, handler, listener);
            case C.TYPE_SS:
                return new SsMediaSource.Factory(
                        new DefaultSsChunkSource.Factory(mediaDataSourceFactory), manifestDataSourceFactory)
                        .createMediaSource(uri, handler, listener);
            case C.TYPE_HLS:
                return new HlsMediaSource.Factory(mediaDataSourceFactory)
                        .createMediaSource(uri, handler, listener);
            case C.TYPE_OTHER:
                return new ExtractorMediaSource.Factory(mediaDataSourceFactory)
                        .createMediaSource(uri, handler, listener);
            default:
                throw new IllegalStateException("Unsupported type: " + type);
        }
    }

    @Override
    public void loadPreview(long currentPosition, long max) {
        player.setPlayWhenReady(false);
        GlideApp.with(imageView)
                .load(thumbnailsUrl)
                .override(Target.SIZE_ORIGINAL, Target.SIZE_ORIGINAL)
                .transform(new GlideThumbnailTransformationPB(currentPosition))
                .into(imageView);

    }
}

File: app/src/main/java/vn/loitp/a/cv/video/exo/mng/PlayerManager.kt
package vn.loitp.a.cv.video.exo.mng

import android.app.Activity
import android.content.Context
import android.net.Uri
import android.view.ViewGroup
import android.widget.ImageButton
import com.google.android.exoplayer2.*
import com.google.android.exoplayer2.C.ContentType
import com.google.android.exoplayer2.ext.ima.ImaAdsLoader
import com.google.android.exoplayer2.source.MediaSource
import com.google.android.exoplayer2.source.ProgressiveMediaSource
import com.google.android.exoplayer2.source.TrackGroupArray
import com.google.android.exoplayer2.source.ads.AdsMediaSource
import com.google.android.exoplayer2.source.dash.DashMediaSource
import com.google.android.exoplayer2.source.hls.HlsMediaSource
import com.google.android.exoplayer2.source.smoothstreaming.SsMediaSource
import com.google.android.exoplayer2.trackselection.AdaptiveTrackSelection
import com.google.android.exoplayer2.trackselection.DefaultTrackSelector
import com.google.android.exoplayer2.trackselection.TrackSelectionArray
import com.google.android.exoplayer2.ui.PlayerView
import com.google.android.exoplayer2.upstream.DataSource
import com.google.android.exoplayer2.upstream.DefaultDataSourceFactory
import com.google.android.exoplayer2.util.Util
import com.google.android.exoplayer2.video.VideoListener
import com.loitp.R
import com.loitp.core.ext.*
import com.loitp.core.utils.AppUtils

/**
 * Created by Loitp on 04,August,2022
 * Galaxy One company,
 * Vietnam
 * +840766040293
 * freuss47@gmail.com
 */
class PlayerManager : AdsMediaSource.MediaSourceFactory {
    private var adsLoader: ImaAdsLoader? = null
    private var dataSourceFactory: DataSource.Factory? = null
    var player: SimpleExoPlayer? = null
    private var screenW: Int = 0

    var videoW = 0
    var videoH = 0

    @Suppress("unused")
    val contentPosition: Long
        get() = player?.contentPosition ?: 0

    val duration: Long
        get() = player?.duration ?: 0

    constructor(context: Context) {
        this.adsLoader = null
        this.screenW = screenWidth
        dataSourceFactory = DefaultDataSourceFactory(context, AppUtils.appName)
    }

    constructor(context: Context, urlIMAAd: String) {
        if (urlIMAAd.isNotEmpty()) {
            adsLoader = ImaAdsLoader(context, Uri.parse(urlIMAAd))
        }
        this.screenW = screenWidth
        dataSourceFactory = DefaultDataSourceFactory(context, AppUtils.appName)
    }

    @JvmOverloads
    fun init(
        context: Context,
        playerView: PlayerView,
        linkPlay: String,
        contentPosition: Long = 0
    ) {
        if (linkPlay.isEmpty()) {
            throw IllegalArgumentException("linkPlay is empty")
        }
        playerView.controllerShowTimeoutMs = 8000
        playerView.controllerHideOnTouch = true

        // Create a default track selector.
        val videoTrackSelectionFactory = AdaptiveTrackSelection.Factory()
        val trackSelector = DefaultTrackSelector(videoTrackSelectionFactory)

        // Create a player instance.
        player = ExoPlayerFactory.newSimpleInstance(context, trackSelector)

        // Bind the player to the view.
        playerView.player = player

        // This is the MediaSource representing the content media (i.e. not the ad).
        // String contentUrl = context.getString(R.string.content_url);
        val contentMediaSource = buildMediaSource(Uri.parse(linkPlay))

        // Compose the content media source into a new AdsMediaSource with both ads and content.
        var mediaSourceWithAds: MediaSource? = null
        adsLoader?.let {
            it.setPlayer(player)
            mediaSourceWithAds = AdsMediaSource(
                contentMediaSource,
                this,
                it,
                playerView
            )
        }

        player?.let {
            // Prepare the player with the source.
            if (mediaSourceWithAds == null) {
                it.prepare(contentMediaSource)
            } else {
                it.prepare(mediaSourceWithAds)
            }
            it.playWhenReady = true
            it.seekTo(contentPosition)
            it.addListener(object : Player.EventListener {
                override fun onTracksChanged(
                    trackGroups: TrackGroupArray?,
                    trackSelections: TrackSelectionArray?
                ) {
                }

                override fun onLoadingChanged(isLoading: Boolean) {
                }

                override fun onPlayerStateChanged(playWhenReady: Boolean, playbackState: Int) {
                    when (playbackState) {
                        Player.STATE_BUFFERING -> {
                            if (playerView.controllerShowTimeoutMs == 0) {
                                playerView.controllerShowTimeoutMs = 8000
                            }
                            if (!playerView.controllerHideOnTouch) {
                                playerView.controllerHideOnTouch = true
                            }
                        }
                        Player.STATE_IDLE -> {
                        }
                        Player.STATE_READY -> {
                        }
                        Player.STATE_ENDED -> {
                            playerView.apply {
                                showController()
                                controllerShowTimeoutMs = 0
                                controllerHideOnTouch = false
                            }
                        }
                    }
                }

                override fun onPlayerError(error: ExoPlaybackException?) {}

                override fun onPlaybackParametersChanged(playbackParameters: PlaybackParameters?) {}
            })

            it.addVideoListener(object : VideoListener {
                override fun onVideoSizeChanged(
                    width: Int,
                    height: Int,
                    unappliedRotationDegrees: Int,
                    pixelWidthHeightRatio: Float
                ) {
                    videoW = width
                    videoH = height
                }

                override fun onSurfaceSizeChanged(width: Int, height: Int) {
                }

                override fun onRenderedFirstFrame() {}
            })
        }
    }

    fun reset() {
        player?.let {
            it.release()
            player = null
        }
    }

    fun release() {
        player?.let {
            it.release()
            player = null
        }
        adsLoader?.release()
    }

    // AdsMediaSource.MediaSourceFactory implementation.
    override fun createMediaSource(uri: Uri): MediaSource {
        return buildMediaSource(uri)
    }

    override fun getSupportedTypes(): IntArray {
        // IMA does not support Smooth Streaming ads.
        return intArrayOf(C.TYPE_DASH, C.TYPE_HLS, C.TYPE_OTHER)
    }

    private fun buildMediaSource(uri: Uri): MediaSource {
        return when (@ContentType val type = Util.inferContentType(uri)) {
            C.TYPE_DASH -> DashMediaSource.Factory(dataSourceFactory).createMediaSource(uri)
            C.TYPE_SS -> SsMediaSource.Factory(dataSourceFactory).createMediaSource(uri)
            C.TYPE_HLS -> HlsMediaSource.Factory(dataSourceFactory).createMediaSource(uri)
            C.TYPE_OTHER -> ProgressiveMediaSource.Factory(dataSourceFactory).createMediaSource(uri)
            else -> throw IllegalStateException("Unsupported type: $type")
        }
    }

    fun toggleFullscreen(activity: Activity) {
        if (activity.isLandscape()) {
            // land -> port
            activity.toggleFullscreen(isFullScreen = false)
            activity.changeScreenPortrait()
        } else {
            // port -> land
            activity.toggleFullscreen(isFullScreen = true)
            activity.changeScreenLandscape()
        }
    }

    // for other sample not UZVideo
    fun updateSizePlayerView(
        playerView: PlayerView,
        exoFullscreen: ImageButton
    ) {
        if (playerView.context.isLandscape()) {
            playerView.layoutParams.width = ViewGroup.LayoutParams.MATCH_PARENT
            playerView.layoutParams.height = ViewGroup.LayoutParams.MATCH_PARENT
            exoFullscreen.setImageResource(R.drawable.player_collapse)
        } else {
            playerView.layoutParams.height = screenW * 9 / 16
            exoFullscreen.setImageResource(R.drawable.player_fullscreen)
        }
        playerView.requestLayout()
    }

    fun pauseVideo() {
        player?.playWhenReady = false
    }

    fun resumeVideo() {
        player?.playWhenReady = true
    }

    @Suppress("unused")
    fun seekTo(mls: Long) {
        player?.seekTo(mls)
    }
}
